stdlib/SuiteSparse/src/cholmod.jl

```diff
@@ -973,60 +973,53 @@ function Vector{T}(D::Dense{T}) where T
+        increment(unsafe_wrap(Array, s.i, (s.nzmax,), own = false)),
+        copy(unsafe_wrap(Array, s.x, (s.nzmax,), own = false)))
+end
+
+# Trim extra elements in rowval and nzval left around sometimes by CHOLMOD rutines
+function _trim_nz_builder!(m, n, colptr, rowval, nzval)
+    l = colptr[end] - 1
+    resize!(rowval, l)
+    resize!(nzval, l)
+    SparseMatrixCSC(m, n, colptr, rowval, nzval)
+end
+
 function SparseMatrixCSC{Tv,SuiteSparse_long}(A::Sparse{Tv}) where Tv
     s = unsafe_load(pointer(A))
     if s.stype != 0
         throw(ArgumentError("matrix has stype != 0. Convert to matrix " *
             "with stype == 0 before converting to SparseMatrixCSC"))
     end
-
-    B = SparseMatrixCSC(s.nrow, s.ncol,
-        increment(unsafe_wrap(Array, s.p, (s.ncol + 1,), own = false)),
-        increment(unsafe_wrap(Array, s.i, (s.nzmax,), own = false)),
-        copy(unsafe_wrap(Array, s.x, (s.nzmax,), own = false)))
-
-    if s.sorted == 0
-        return SparseArrays.sortSparseMatrixCSC!(B)
-    else
-        return B
-    end
+    args = _extract_args(s)
+    s.sorted == 0 && sortBuffers!(args...);
+    return _trim_nz_builder!(args...)
 end
 
-function (::Type{Symmetric{Float64,SparseMatrixCSC{Float64,SuiteSparse_long}}})(A::Sparse{Float64})
+function Symmetric{Float64,SparseMatrixCSC{Float64,SuiteSparse_long}}(A::Sparse{Float64})
     s = unsafe_load(pointer(A))
-    if !issymmetric(A)
-        throw(ArgumentError("matrix is not symmetric"))
-    end
-
-    B = Symmetric(SparseMatrixCSC(s.nrow, s.ncol,
-        increment(unsafe_wrap(Array, s.p, (s.ncol + 1,), own = false)),
-        increment(unsafe_wrap(Array, s.i, (s.nzmax,), own = false)),
-        copy(unsafe_wrap(Array, s.x, (s.nzmax,), own = false))), s.stype > 0 ? :U : :L)
-
+    issymmetric(A) || throw(ArgumentError("matrix is not symmetric"))
+    args = _extract_args(s)
     if s.sorted == 0
-        return SparseArrays.sortSparseMatrixCSC!(B.data)
-    else
-        return B
+        sortBuffers!(args...)
+        return _trim_nz_builder!(args...)
     end
+    return Symmetric(_trim_nz_builder!(args...), s.stype > 0 ? :U : :L)
 end
 convert(T::Type{Symmetric{Float64,SparseMatrixCSC{Float64,SuiteSparse_long}}}, A::Sparse{Float64}) = T(A)
 
 function Hermitian{Tv,SparseMatrixCSC{Tv,SuiteSparse_long}}(A::Sparse{Tv}) where Tv<:VTypes
     s = unsafe_load(pointer(A))
-    if !ishermitian(A)
-        throw(ArgumentError("matrix is not Hermitian"))
-    end
-
-    B = Hermitian(SparseMatrixCSC(s.nrow, s.ncol,
-        increment(unsafe_wrap(Array, s.p, (s.ncol + 1,), own = false)),
-        increment(unsafe_wrap(Array, s.i, (s.nzmax,), own = false)),
-        copy(unsafe_wrap(Array, s.x, (s.nzmax,), own = false))), s.stype > 0 ? :U : :L)
-
+    ishermitian(A) || throw(ArgumentError("matrix is not symmetric"))
+    args = _extract_args(s)
     if s.sorted == 0
-        return SparseArrays.sortSparseMatrixCSC!(B.data)
-    else
-        return B
+        @assert false
```

Comment:
I was meaning to ask. I can't say I understand this cholmod stuff perfectly, but: the case with `sorted==0` here makes `Hermitian(::Sparse)` return a `SparseArrayCSC` instead of an `Hermitian`. Was this intended? This branch of the code may be dead, as tests pass with `@assert false`. I will take the assert out of course, but can someone clarify?